docs: update blog (#41177)
ant-design/ant-design: docs/blog/contributor-development-maintenance-guide.en-US.md

---
title: Contributor development maintenance guide
date: 2023-03-10
author: kiner-tang
---

> Hi, my name is [kiner-tang](https://github.com/kiner-tang). My personal work is closely related to Ant Design, so I can be considered a heavy user of Ant Design.It is for this reason that I am slowly working my way up from a user to a Contributor, contributing some of the problems I encountered and the new features I concluded to the community, and finally being honored to become a member of Ant Design Collaborator.

Reviewer: Hi, I'm [kiner-tang](https://github.com/kiner-tang) who is a heavy user of Ant Design. As the same reason, I come up to be a contributor for fixing bug and some features. Finally, it's my honer to become a member of Ant Design Collaborator.

Author: done

> During the journey from user to contributor, and then from contributor to collaborator, I also encountered many problems. I would like to take this opportunity to sort out and summarize these problems, hoping to be helpful to new contributors and collaborators joining Ant Design community.

Reviewer: ... I would like to share my experience with the problem I meet. Hope it can help you to join the Ant Design community.

Author: done

## Some frequently asked Questions

### About dependent version

> Popular package managers, whether npm, yarn, or pnpm, offer version-locking solutions that, most of the time, help us avoid version-inconsistencies in packages.However, in Ant Design projects, many functions depend on the original components in the `react-component` repository. We expect that when we encounter some bugs, after the `react-component` fixes and releases the patch version, Instead of manually upgrading the version in the Ant Design project, we can install the latest patch installation package only by reinstalling the dependencies.At this point, the version-lock file provided by the package manager is the main obstacle to automatic updates, because once the version-lock file is available, reinstalling the dependency will also install the version specified by the lock file and cannot be upgraded to the patch version.

Reviewer: There are various of package mangers, such as ..., providing version-locking solutions to avoid version-inconsistencies issue.

Reviewer: many functions depend on the underlying dependencies of `react-component`.

Reviewer: We hope Ant Design follow the rule of semver. When underlying dependencies fix the bug, user can easily reinstall the `node_modules` to fix instead care about too many detail about internal packages.

Author: done

Based on the above reasons, we adopt the following methods:

1. Add `package-lock.json`, `yarn.lock`, `pnpm-lock.yaml` and other version lock files to `.gitignore` without version tracking.

2. In `package.json`, for the dependencies that we want to update when there is a new patch version, we use `~` description version number to allow patch version update.

```json
{
  "dependencies": {
    "rc-cascader": "~3.9.0"
  }
}
```

For the difference between `^` and `~` in the version description of package.json, see [What's the difference between tilde(~) and caret(^) in package.json](https://stackoverflow.com/questions/22343224/whats-the-difference-between-tilde-and-caret-in-package-json).

Thus, when our dependency such as rc-cascader fixes a bug and releases a patch version such as: `3.9.1`, then the user's latest installed version is `3.9.1`. For the maintainer of Ant Design, we only need to execute the following command:

```bash
git clean -fdx
npm i
```

### Snapshot update problem

> In Ant Design, we use Jest for unit test. Many first-time contributors of Ant Design may be confused that CI failed after opening a pull request in which they just modify the text in a demo. This starts with Ant Design's snapshot test.

Reviewer: > This starts with Ant Design's snapshot test.

This is caused by the snapshot diff test.

Reviewer: 看中文的时候没感觉，看英文的时候发现 Ant Design 有点多

> In Ant Design, we use jest for unit testing and coverage testing, with many first-time references Students of Ant Design project development may often find that they just modify the text in a demo. Why does the CI test task fail as soon as they push it? This starts with Ant Design's snapshot detection.

Reviewer 2: ```suggestion
In Ant Design, we use Jest for unit test. Many first-time contributors of Ant Design may be confused that CI failed after opening a pull request in which they just modify the text in a demo. This starts with Ant Design's snapshot test.
```

> In most libraries, the concept of idempotence is emphasized, meaning that the same method, no matter how many times I execute the line, as long as the input parameters are the same, then the result will be the same.In Ant Design, I think the biggest use of snapshot detection is to verify the idempotence of our demo, so as to ensure the stability and certainty of component output.In fact, the principle of snapshot detection is very simple, is to save our demo generated html string, the next time in the running of the test task for comparison, if there is a difference that indicates that the snapshot detection failed.

Reviewer 2: ```suggestion
In most libraries, idempotence is emphasized. With one method and same parameters, the result should be the same no matter how many times it is executed. In Ant Design, the main use of snapshot test is to verify the idempotence of demos, in order to ensure the stability and certainty of components. Actually, snapshot test is simple. By comparing HTML string generated by each demo in different runnings, we could get which one is broken in PR.
```

Going back to the original question, what should we do if we change a demo and find that the snapshot detection fails?

> 1. First of all, we need to check the inconsistency of snapshot comparison to see if it is the expected change. If only the text is changed, and only the text is different from the snapshot comparison result, we just need to run the following command to update the snapshot:

Reviewer 2: ```suggestion
1. First, we need to check the snapshot diff to see whether it is the expected change. If there is only the text you modified in demo differs from the origin one, we just need to run the following command to update the snapshot:
```

   ```bash
   # Run the test command to update snapshot
   npm run test -u
   ```

> 2. However, if the changes during snapshot comparison exceed the scope of your modification this time, for example, you only changed the text, but found that classnames in snapshots have changed, which obviously does not meet expectations, we need to rectify the causes.

Reviewer 2: ```suggestion
2. However, we should find what's going wrong if the diff of snapshot is not only cause by your modification this time. For example, you only modified the text, but found that classnames in snapshots have changed, which obviously does not meet expectations.
```

   The common causes are as follows:

>    - Local dependency is too old, it is possible to pull the latest code, but did not update the local dependency, resulting in the dependency package version is too low and the output results are inconsistent.The

Reviewer 2: ```suggestion
   - Local dependency is out of date. This may happen when you pull the latest code, but did not update the local dependency. Deleting `lock` file, `node_modules`, and then reinstalling dependencies could solve this problem.
```

>      solution is as simple as deleting the lock file、node_modules，and reinstall dependencies.

Reviewer 2: ```suggestion
```

>    - You may have locally changed the core code in addition to the demo, causing the logic to change. You need to go over this change carefully.

Reviewer 2: ```suggestion
   - You may have locally changed the source code not only in demos,  which would cause unexpected change. You need to go through your change carefully.
```

### rc-x Library dependencies

> In Ant Design, most components are an upper encapsulation of a component based on `react-component`. Therefore, if a user reports a bug, when we troubleshoot the problem, if we find the problem is in the `@rc-component/xxx` or `rc-xxx` component, Then we need to put pr on those components to fix it.After the fix, we need to verify the fix results in the Ant Design project, then we can link the project to the Ant Design project for verification.For example:

Reviewer 2: ```suggestion
In Ant Design, most components are an upper encapsulation of a component based on `react-component`. Therefore, if we find a problem of `@rc-component/xxx` or `rc-xxx` component when troubleshooting, we need to open PR on those repositories to fix it. After coding, we need to verify that the problem in Ant Design is resolved, in which case we can link the project to Ant Design for verification. For example:
```

> Run `npm link` under the rc project.

Reviewer 2: ```suggestion
Run `npm link` in the repo you are working on.
```

![image](https://user-images.githubusercontent.com/10286961/224603053-98488c2d-f33c-4c25-8c09-6c790cfcdbf6.png)

Run `npm link "Project name"` in Ant Design

![image](https://user-images.githubusercontent.com/10286961/224603065-95715727-83d0-4ef9-81e4-3b7065aaf73e.png)

> Once we have verified, we can mention pr to the rc component.It should be noted that link may cause an exception when running the test command. Therefore, after local verification, we need to run the following

Reviewer 2: ```suggestion
Once we have verified, we can open PR to the repo. Noted that link may cause exceptions when running the test command. Therefore, we need to run the following commands locally to delete the package from link.
```

> command locally to delete the package from link:

Reviewer 2: ```suggestion
```

```bash
npm unlink "rc-field-form" --no-save
npm i
```

> When the pr is finally merged, usually the maintainer will release a version. If the patch version is released, then you only need to install and verify it in the Ant Design project, but if the minor version is released, Then we also need to take the initiative to upgrade the version in the Ant Design project, and after passing the local verification, a separate pr to Ant Design for upgrade repair.

Reviewer 2: ```suggestion
When the PR is finally merged, usually the maintainers will release a version. If the patch version is released, you only need to install and verify it in Ant Design. But if the minor version is released, you need to upgrade it in Ant Design. After local verification, a separate PR to Ant Design could be opened for bugfix.
```

> ### Property expiration

Reviewer 2: ```suggestion
### Deprecation
```

> In a large project, if you want to scrap an property or method, it is actually very troublesome, because your project may already have a lot of items in use, if you kill it, then others upgrade the console screen will be red, or even can't run.But in the process of project iteration, we will encounter more and more scenes and problems, a long time ago solution may no longer fit, really need to scrap, then, we need to adopt a soft, less radical way to scrap, leave enough time for users to modify.

Reviewer 2: ```suggestion
In a large project, if you want to remove an property or a method, it is actually very troublesome. Since your project may already have a lot of items in use, other projects depend on it may get broken. But as the project iterated, we will encounter more and more problems which solutions long time ago may no longer fit. Then, we need to adopt a soft, less radical way to remove it, leaving enough time for users to modify.
```

> In Ant Design, we adopt a **five-step strategic** obsolescence property or method:

Reviewer 2: ```suggestion
In Ant Design, we adopt a **five-step strategic** to deprecate property or method:
```

> 1. **Add an expiration mark to the property**

Reviewer 2: ```suggestion
1. **Add deprecated tag to the property**
```

   ![003](https://user-images.githubusercontent.com/10286961/224358324-8f72f2c0-d5bb-4281-9b29-7e2428353449.png)

>    When we add the above expired tag, we can see later using this variable:

Reviewer 2: ```suggestion
   After adding the above deprecated tag, we can see warning when using this property:
```

   ![004](https://user-images.githubusercontent.com/10286961/224358351-958a168d-41de-44b0-8244-2f8d67c4d13a.png)

2. **Add console warnings**

![005](https://user-images.githubusercontent.com/10286961/224358371-09f08f79-8c95-4126-b382-59311bb702d6.png)

> It is important to note that after adding console warnings, we need to add a test case to the test case to test whether the warnings will be displayed if the expired properties are used, so as to ensure that the warnings can be displayed normally.

Reviewer 2: ```suggestion
It is important that after adding console warnings, we need to add a test case to test whether the warnings will be displayed when the deprecated properties are used.
```

![006](https://user-images.githubusercontent.com/10286961/224358407-3d89d2f5-b4aa-48b4-aab8-1331a0f620fa.png)

> 3. **Update document**: The above two steps are mainly to warn developers in the editor and browser preview. Next, we need to update the relevant software documentation. Generally speaking, if we need an expired property, it means that it is not recommended for users to use. And add the description of the new property simultaneously.(PS: If it cannot be deleted due to special circumstances, please specify the version supported by this property and the alternative scheme in the remarks column)

Reviewer 2: ```suggestion
3. **Update document**: The above two steps are mainly to warn developers in the editor and browser. Next, we need to update the related document. Generally speaking, if we have a property deprecated, it means that it is not recommended for users to use. So we need to remove the related document and add the description of the new property at the mean time. If it cannot be deleted due to special circumstances, please specify the version this property is supported and the alternative scheme in the remarks column.
```

> 4. **Compatibility between old and new**: Now that we're done with the early tips, it's important to make sure that the new property and the old one live in harmony for a while until the old one dies. That is to say, currently we use expired property to work normally, in addition to the warning, can not have any changes.

Reviewer 2: ```suggestion
4. **Compatibility between old and new**: Now that we're done with the above tips, it's important to make sure that the new property and the deprecated one both works until the deprecated one is removed. That is to say, we must have the deprecated property work normally. We should not apply any changes in addition to the warning.
```

   ![007](https://user-images.githubusercontent.com/10286961/224358439-76c42c78-e244-42bd-8935-b08f536931a2.png)

>    We can do something like this, making sure that the new property takes precedence over the old one, and only trying to use the old one if no new one exists.At this point, our property expiration work is done.

Reviewer 2: ```suggestion
   We can do something like this to make sure that the new property takes precedence over the old one. At this point, our property deprecation work is done.
```

> 5. **Wipe out**: After our properties have expired for a certain period of time, usually when the major version number is upgraded, we can remove the obsolescence. We also need to delete comments, warnings, test cases, and documents that were added to obsolete the property etc. At this point our property abandonment work is done.

Reviewer 2: ```suggestion
5. **Remove**: After our properties have been deprecated for a certain period of time, usually when the major version is released, we can remove the them. We also need to delete comments, warnings, test cases, and documents that were related the property. At this point our property removal work is done.
```

## Take it to the next level

> After contributing for some time, I believe you have gained a deeper understanding of Ant Design's overall development process and project architecture. At this point, you may want to take on more tasks and contribute more to the open source community.Then, you can apply to be a Collaborator at Ant Design by commenting on the link below for more active contributors. [Add Collaborator permission for some active contributors](https://github.com/ant-design/ant-design/issues/3222), The Collaborators will then start the voting process, and after passing the voting, they will invite you to officially become a Collaborator at Ant Design.

Reviewer 2: ```suggestion
After contributing for some time, we believe you have gained a deeper understanding of Ant Design's overall development process and project architecture. At this point, you may want to take on more tasks and contribute more to the open source community. Then, you can apply to be a Collaborator at Ant Design by commenting on the link below for more active contributors. [Add Collaborator permission for some active contributors](https://github.com/ant-design/ant-design/issues/3222), The Collaborators will then start the voting process, and after passing the voting, they will invite you to officially become a Collaborator at Ant Design.
```

After you become a newly appointed Ant Design collaborator, you will be assigned some additional rights, such as:

- Tag the issue
- Close the issue
- Free to create branches under the Ant Design project
- Free to merge approved PR

- ...

There are some caveats to this.

### Merge PR

At this point, you can merge the approved PR, but merging code is tricky, knowing when to use squash and when not to.

![008](https://user-images.githubusercontent.com/10286961/224358476-2332e36f-0adf-486f-8b17-1b2ad34926aa.jpg)

> - **Merge pull request**: Normally merge the current PR into the target branch usually when branch to branch merge each other, do not merge the commit record, do not merge PR, or you will forget to add the target branch too many redundant commit records.

Reviewer: Merge the current PR into the target branch normally (usually used when branches merge with each other, and commit records will not be merged. Do not use this when merging PRs, otherwise too many redundant commit records will be added to the target branch)

> - **Squash and merge**: Merge commit records into a squash and merge into the target branch(this is usually used for merging PR).

Reviewer: Merge the commit record into one and merge it into the target branch (this is usually used for merging PRs)

- **Rebase and merge**：When you want to adjust the baseline

> Before merging branches, it is a good idea to check with the author that they have been fully modified and that they have been carefully reviewed. And at least one maintainer approves this PR. Determine which of the above scenarios this PR is, and then choose the appropriate way to merge.

Reviewer: Merge in Ant Design requires at least one collaborator's approve. Collaborator who approved should ensure PR are fully reviewed and confirm.

### Identify XY Problems

> As a collaborator, you may often communicate with users, so you are very likely to encounter XY problems, XY problems seriously affect our troubleshooting efficiency, therefore, we need to have a basic cognition and discrimination of XY problems, and learn how to avoid falling into the trap of XY problems.

Reviewer: 英文一读，发现这个有点和 Collaborator 关系不大，不是 Collaborator 沟通也是需要理解问题。可以不需要 `As a collaborator`：

During routine maintenance, some issues can easily reproduce but some others are not. Report sometime get stuck in XY problem which makes requirement or bug question strange to understand. Thus, before resolving the superficial problems, we should dig into the real question.

- [X-Y PROBLEM](https://xyproblem.info/)

## Epilogue

> Finally, we sincerely hope that more and more contributors and collaborators can join us to build a more efficient and elegant Ant Design.

Reviewer: 一般不会特地加 `Finally`:

Hoping that more and more contributors and collaborators will join us to build a more efficient and elegant Ant Design.

---
title: Contributor development maintenance guide
date: 2023-03-10
author: kiner-tang
---

> Hi, my name is [kiner-tang(文辉)](https://github.com/kiner-tang). My personal work is closely related to AntDesign, so I can be considered a heavy user of AntDesign.It is for this reason that I am slowly working my way up from a user to a Contributor, contributing some of the problems I encountered and the new features I concluded to the community, and finally being honored to become a member of AntDesign Collaborator.

Reviewer: 英文版里不用出现中文

Based on the above reasons, we adopt the following methods:

1. Add `package-lock.json`, `yarn.lock`, `pnpm-lock.yaml` and other version lock files to `.gitignore` without version tracking.

2. In `package.json`, for the dependencies that we want to update when there is a new patch version, we use `~` description version number to allow patch version update.

```json
{
  "dependencies": {
    "rc-cascader": "~3.9.0"
  }
}
```

> For the difference between `^` and `~` in the version description of package.json, see [package.json 文件^和~区别](https://blog.csdn.net/shan1991fei/article/details/125430971).

Reviewer: 这里需要找一下英文版对应的文档链接，不能链接到中文的blog

Author: ok

![008](https://user-images.githubusercontent.com/10286961/224358476-2332e36f-0adf-486f-8b17-1b2ad34926aa.jpg)

- **Merge pull Request**: Normally merge the current PR into the target branch usually when branch to branch merge each other, do not merge the commit record, do not merge PR, or you will forget to add the target branch too many redundant commit records.
- **Squash and merge**: Merge commit records into a squash and merge into the target branch(this is usually used for merging PR).
- **Rebase and merge**：When you want to adjust the baseline

Before merging branches, it is a good idea to check with the author that they have been fully modified and that they have been carefully reviewed. And at least one maintainer approves this PR. Determine which of the above scenarios this PR is, and then choose the appropriate way to merge.

### Identify XY Problems

As a collaborator, you may often communicate with users, so you are very likely to encounter XY problems, XY problems seriously affect our troubleshooting efficiency, therefore, we need to have a basic cognition and discrimination of XY problems, and learn how to avoid falling into the trap of XY problems.

> - [高效沟通的拦路虎——XY 问题](https://zhuanlan.zhihu.com/p/612569093)

Reviewer: 同理，英文文章不要连到中文 blog

Author: ok
